node/consensus/src/transactions_queue.rs

```diff
@@ -0,0 +1,188 @@
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    cmp::Reverse,
+    collections::{BTreeMap, HashMap, hash_map::Entry},
+    num::NonZeroUsize,
+};
+
+use anyhow::{Result, bail};
+use lru::LruCache;
+use snarkvm::{ledger::Transaction, prelude::*};
+
+use crate::{CAPACITY_FOR_DEPLOYMENTS, CAPACITY_FOR_EXECUTIONS};
+
+pub struct TransactionsQueue<N: Network> {
+    pub deployments: TransactionsQueueInner<N>,
+    pub executions: TransactionsQueueInner<N>,
+}
+
+impl<N: Network> Default for TransactionsQueue<N> {
+    fn default() -> Self {
+        Self {
+            deployments: TransactionsQueueInner::new(CAPACITY_FOR_DEPLOYMENTS),
+            executions: TransactionsQueueInner::new(CAPACITY_FOR_EXECUTIONS),
+        }
+    }
+}
+
+impl<N: Network> TransactionsQueue<N> {
+    pub fn contains(&self, transaction_id: &N::TransactionID) -> bool {
+        self.executions.contains(transaction_id) || self.deployments.contains(transaction_id)
+    }
+
+    pub fn insert(&mut self, transaction_id: N::TransactionID, transaction: Transaction<N>) -> Result<()> {
+        if transaction.is_execute() {
+            self.executions.insert(transaction_id, transaction)
+        } else {
+            self.deployments.insert(transaction_id, transaction)
+        }
+    }
+
+    pub fn transactions(&self) -> impl Iterator<Item = (N::TransactionID, Transaction<N>)> {
+        self.deployments
+            .priority_queue
+            .transactions
+            .clone()
+            .into_iter()
+            .chain(self.deployments.queue.clone())
+            .chain(self.executions.priority_queue.transactions.clone())
+            .chain(self.executions.queue.clone())
+    }
+}
+
+pub struct TransactionsQueueInner<N: Network> {
+    capacity: usize,
+    queue: LruCache<N::TransactionID, Transaction<N>>,
+    priority_queue: PriorityQueue<N>,
+}
+
+impl<N: Network> TransactionsQueueInner<N> {
+    fn new(capacity: usize) -> Self {
+        Self {
+            capacity,
+            queue: LruCache::new(NonZeroUsize::new(capacity).unwrap()),
+            priority_queue: Default::default(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.queue.len().saturating_add(self.priority_queue.len())
+    }
+
+    fn contains(&self, transaction_id: &N::TransactionID) -> bool {
+        self.queue.contains(transaction_id) || self.priority_queue.transactions.contains_key(transaction_id)
+    }
+
+    fn insert(&mut self, transaction_id: N::TransactionID, transaction: Transaction<N>) -> Result<()> {
+        let priority_fee = transaction.priority_fee_amount()?;
```

Comment:
Cleanup nit: you're only using `let has_priority_fee = transaction.priority_fee_amount()?.is_zero()` below.